---
layout: post
title:  BooleaBayes Overview
description: This analysis was written to analyze transcription factor networks for Small Cell Lung Cancer phenotypes. I've written a primer on gene regulatory network dynamics and how our computational tool is able to define them and make predictions about the future using gene expression data-- RNA sequencing data that tells you the identity of a population of cells.
date: 2021-01-31
categories: booleabayes
tags: code blog writing
---

To see the code and documentation, please visit <a href= 'https://github.com/smgroves/BooleaBayes'>Github</a> or <a href='https://pypi.org/project/booleabayes/'>PYPI</a> for instructions on installing the package.

## Part 1: The Why
In this first post, I explain why we care about transcription factor networks and how they can help us treat Small Cell Lung Cancer. \
<a href='https://smgroves.github.io/blog/2021/booleabayes1/'>Read More</a>.

## Part 2: Network structure and dynamics
This post takes a look what information we will need to gather in order to understand how cell identity is controlled. \
<a href='https://smgroves.github.io/blog/2021/booleabayes2/'>Read More</a>.

## Part 3: Using data to build a network for Small Cell Lung Cancer
We begin to look at how BooleaBayes, the computational tool developed in the Quaranta lab by David Wooten, PhD and me, can be applied to Small Cell Lung Cancer Data. \
<a href='https://smgroves.github.io/blog/2021/booleabayes3/'>Read More</a>.

## Part 4: How good is our model?
How can we be sure that our model's predictions will be accurate and tell us something new?

## Part 5: So... can we cure Small Cell Lung Cancer yet?
Where we are with targeting cell plasticity as a form of therapy.

*Posts are adapted from my talk on <a href = "http://smgroves.github.io/about/talks/">Network Control of Phenotypic Plasticity in SCLC.</a>
