---
layout: post
title: 'Vis for Deep Learning for Biology'
date: 2017-05-16
tags: research
categories: research
tabs: true
image: anl_lcfbuilding.png-srcw.jpg
---

## Vis for Deep Learning for Biology


Argonne Leadership Computing Facility

![image](/images/anl_lcfbuilding.png-srcw.jpg){:style="max-width: 100%"}

Deep learning is often referred to as a &lsquo;black box&rsquo;, because, aside from selecting model and the model hyperparameters (like the learning rate, activation function parameters), users don&rsquo;t have access to what happens inside the model.  We can know that the model works by measuring its accuracy on some testing data set, suggesting that the model must have &lsquo;learned&rsquo; something meaningful from the training data.  But how can we peek inside and &lsquo;see&rsquo; what the model &lsquo;sees&rsquo;?<br><br>
Biologists have recently begun to explore how deep learning approaches could be used to address questions involving diverse biological datasets.  For this community, producing an accurate model is important, but also insufficient, since the larger goal is to contribute to our understanding of complex biological systems and phenomena.  Visualization could help biology researchers &lsquo;peak&rsquo; inside the black box and relate deep learning mechanics to existing biological knowledge.  However, most visualization techniques for deep learning have been developed for images and text, so many of the encodings and layouts are designed for input data with fundamentally different perceptual properties than biological data sets, such as gene sequence data or gene expression data.<br><br> 
This research explores current approaches in visualization for deep learning and how these approaches may (or may not) be applied to biological data.  This work is conducted with researchers at Argonne National Labs studying deep learning for cancer research and genomics.
